### PR TITLE
Add convert function

### PR DESCRIPTION
**Pull request by gongweibao**
Fix [convert library](https://github.com/PaddlePaddle/Paddle/blob/develop/doc/design/cluster_train/data_dispatch.md#%E8%BD%AC%E6%8D%A2%E5%BA%93)

### CODE REVIEW

File: python/paddle/v2/dataset/common.py [hunk 1]
```diff
@@ -149,3 +149,47 @@ def reader():
                     yield line
 
     return reader
+
+
+def convert(output_path, eader, num_shards, name_prefix):
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: eader -> reader

File: python/paddle/v2/dataset/common.py [hunk 2]
```diff
@@ -149,3 +149,47 @@ def reader():
...
+    Convert data from reader to recordio format files.
+
+    :param output_path: directory in which output files will be saved.
+    :param reader: a data reader, from which the convert program will read data instances.
+    :param num_shards: the number of shards that the dataset will be partitioned into.
+    :param name_prefix: the name prefix of generated files.
+    """
+
+    def open_needs(idx):
+        n = "%s/%s-%05d" % (output_path, name_prefix, idx)
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Should follow the format in the design doc:
```text
random_images-00000-of-00099
random_images-00001-of-00099
...
random_images-00099-of-00099
```

File: python/paddle/v2/dataset/common.py [hunk 3]
```diff
@@ -149,3 +149,47 @@ def reader():
...
+    """
+    Convert data from reader to recordio format files.
+
+    :param output_path: directory in which output files will be saved.
+    :param reader: a data reader, from which the convert program will read data instances.
+    :param num_shards: the number of shards that the dataset will be partitioned into.
+    :param name_prefix: the name prefix of generated files.
+    """
+
+    def open_needs(idx):
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: What does `open_needs` mean in English? Maybe a more descriptive name would be better.

File: python/paddle/v2/dataset/common.py [hunk 4]
```diff
@@ -149,3 +149,47 @@ def reader():
...
+        n = "%s/%s-%05d" % (output_path, name_prefix, idx)
+        w = recordio.writer(n)
+        f = open(n, "w")
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Why do we need `f`? `writer.Close` already closes the file it opens.

File: python/paddle/v2/dataset/common.py [hunk 5]
```diff
@@ -149,3 +149,47 @@ def reader():
...
+            w.close()
+
+        if f is not None:
+            f.close()
+
+    idx = 0
+    w = None
+    f = None
+
+    for i, d in enumerate(reader()):
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Sorry I should have mentioned earlier, please consider this issue: https://github.com/PaddlePaddle/Paddle/issues/1915

To randomize, maybe we could have a shuffle_buffer_size as optional parameter. read until the buffer is full, shuffle and then write to RecordIO.

File: python/paddle/v2/dataset/common.py [hunk 6]
```diff
@@ -149,3 +149,47 @@ def reader():
                     yield line
 
     return reader
+
+
+def convert(output_path, eader, num_shards, name_prefix):
+    import recordio
+    import cPickle as pickle
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Please add cPickle installation with fixed version in https://github.com/PaddlePaddle/Paddle/blob/develop/Dockerfile#L55 .

File: python/paddle/v2/dataset/common.py [hunk 7]
```diff
@@ -149,3 +149,47 @@ def reader():
...
+    w = None
+    f = None
+
+    for i, d in enumerate(reader()):
+        if w is None:
+            w, f, idx = open_needs(idx)
+
+        w.write(pickle.dumps(d, pickle.HIGHEST_PROTOCOL))
+
+        if i % num_shards == 0 and i >= num_shards:
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: 这里的逻辑我不是很明白，假设一共有N个shard，这里的逻辑是每N个record，写入下一个shard。不应该是每一个record写入下一个shard吗？

是不是考虑这样：
```go
var writers []writer
// fill writer
writer[i%num_shards].Write(record)
// close all writer once everything is done. Don't close and create a new writer frequently.
```

File: python/paddle/v2/dataset/tests/common_test.py [hunk 8]
```diff
@@ -57,6 +57,22 @@ def test_cluster_file_reader(self):
         for idx, e in enumerate(reader()):
             self.assertEqual(e, str("0"))
 
+    def test_convert(self):
+        def test_reader():
+            def reader():
+                for x in xrange(10):
```
Comment (helinwang) [inline python/paddle/v2/dataset/tests/common_test.py]: I think this test will break when 10 is changed to 4. According to line 191 `if i % num_shards == 0 and i >= num_shards:`

File: python/paddle/v2/dataset/common.py (hunk 6 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Python 2.7 and Python 3.4 include the pickle and cPickle modules already. But I didn't find official answer, there is a [answer](https://askubuntu.com/questions/755431/how-to-install-the-cpickle-module)

File: python/paddle/v2/dataset/common.py (hunk 3 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done

File: python/paddle/v2/dataset/common.py (hunk 2 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done

File: python/paddle/v2/dataset/common.py (hunk 4 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done

File: python/paddle/v2/dataset/common.py (hunk 5 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done

File: python/paddle/v2/dataset/common.py (hunk 7 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done.一开始想错了，把`num_shards`想成每间隔多少个写入一个文件了。汗！

Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: 一开始想如果文件数目比较多，而记录的个数比较少。那样的话会生成空文件。

File: python/paddle/v2/dataset/common.py (hunk 1 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/common.py]: Done

File: python/paddle/v2/dataset/tests/common_test.py (hunk 8 above)
Comment (gongweibao) [inline python/paddle/v2/dataset/tests/common_test.py]: Done

File: python/paddle/v2/dataset/common.py (hunk 6 above)
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Great!

File: python/paddle/v2/dataset/common.py [hunk 9]
```diff
@@ -149,3 +149,57 @@ def reader():
                     yield line
 
     return reader
+
+
+def convert(output_path,
+            reader,
+            num_shards,
+            name_prefix,
+            max_lines_to_shuffle=10000):
```
Comment (helinwang) [inline python/paddle/v2/dataset/common.py]: Maybe 1000 could be better. Say we have images of 200KB per image, 10000 of them is 1.9GB, might be too big in memory consumption.

**Pull request merged**